internal/query/map.go

```diff
@@ -1,81 +1,198 @@
+	return key[:len(key)-2]
 }
 
 // parsePath is an internal function to parse key access path.
 // For example, key[foo][bar] will be parsed to ["foo", "bar"].
-func parsePath(k string, key string) ([]string, error) {
-	rawPath := strings.TrimPrefix(k, key)
-	if rawPath == "" {
-		return nil, fmt.Errorf("expect %s to be a map but got value", key)
+func parsePath(k string) ([]string, error) {
+	firstKeyEnd := strings.IndexByte(k, '[')
+	if firstKeyEnd == -1 {
+		return nil, fmt.Errorf("invalid access to map key %s", k)
 	}
-	splitted := strings.Split(rawPath, "]")
-	paths := make([]string, 0)
-	for _, p := range splitted {
-		if p == "" {
-			continue
-		}
-		if strings.HasPrefix(p, "[") {
+	first, rawPath := k[:firstKeyEnd], k[firstKeyEnd:]
+
+	split := strings.Split(rawPath, "]")
+
+	// Bear in mind that split of the valid map will always have "" as the last element.
+	if split[len(split)-1] != "" {
+		return nil, fmt.Errorf("invalid access to map key %s", k)
+	} else if len(split)-1 > MaxNestedMapDepth {
+		return nil, fmt.Errorf("maximum depth [%d] of nesting in map exceeded [%d]", MaxNestedMapDepth, len(split)-1)
+	}
+
+	// -2 because after split the last element should be empty string.
+	last := len(split) - 2
+
+	paths := []string{first}
+	for i := 0; i <= last; i++ {
+		p := split[i]
+
+		// this way we can handle both error cases: foo] and [foo[bar
+		if strings.LastIndex(p, "[") == 0 {
 			p = p[1:]
 		} else {
 			return nil, fmt.Errorf("invalid access to map key %s", p)
 		}
-		if p == "" {
-			return nil, fmt.Errorf("expect %s to be a map but got array", key)
+		if p == "" && i != last {
+			return nil, fmt.Errorf("unsupported array-like access to map key %s", k)
 		}
 		paths = append(paths, p)
 	}
 	return paths, nil
 }
 
 // setValueOnPath is an internal function to set value a path on dicts.
-func setValueOnPath(dicts map[string]interface{}, paths []string, value []string) {
+func setValueOnPath(dicts map[string]interface{}, paths []string, value []string) error {
 	nesting := len(paths)
+	previousLevel := dicts
 	currentLevel := dicts
 	for i, p := range paths {
 		if isLast(i, nesting) {
-			currentLevel[p] = value[0]
+			if isArrayOnPath(p) {
+				previousLevel[paths[i-1]] = value
+			} else {
+				if _, ok := currentLevel[p]; ok {
+					return fmt.Errorf("trying to set different types at the same key [%s]", p)
+				}
+				currentLevel[p] = value[0]
+			}
 		} else {
-			initNestingIfNotExists(currentLevel, p)
-			currentLevel = currentLevel[p].(map[string]interface{})
+			switch currentLevel[p].(type) {
+			case map[string]any:
+				if isArrayOnPath(paths[i+1]) {
+					return fmt.Errorf("trying to set different types at the same key [%s]", p)
+				}
+			case nil:
+				currentLevel[p] = make(map[string]any)
+			default:
+				return fmt.Errorf("trying to set different types at the same key [%s]", p)
+			}
+			previousLevel = currentLevel
+			currentLevel = currentLevel[p].(map[string]any)
```

Comment:
I suppose that putting this block to a separate function is quite hard because of usage of local variables (e.g. `previousLevel`) but maybe some comment would be nice here 🤔